### ISSUE TITLE
fn:elements-to-maps: `empty`, normalize space ?

### ISSUE THREAD

**Issue by ChristianGruen** [qt4cg/qtspecs#1658]
The current rules say:

> If `empty($EE/(* | text())` (that is, if there are no child elements or text nodes) then: […] `empty`
> If `empty($EE/text()[normalize-space()])` (that is, there are no text node children other than whitespace), then: […] `list`

Maybe it would be consistent to add `[normalize-space()]` to the condition of the `empty` layout?

**Comment by graydon2014** [comment]
I agree that it's more consistent but I'd almost rather have explicit filter rules for the whitespace-only text nodes.

It is often the case that the "no, not that one" text node has `&#x00A0;` in it, rather than XML whitespace; there are a whole bunch of fancy Unicode spaces that show up sometimes and have meaning, too. (Usually between elements, admittedly.) I've wound up using `matches(.,'\P{Zs}')` rather than `normalize-space()` as a test fairly often.

A keep-when option with a regex that does `[matches(.,$keep-when-regex)]` strikes me as preferable to plain normalize-space().

**Comment by michaelhkay** [comment]
>Maybe it would be consistent to add [normalize-space()] to the condition of the empty layout?

The effect of this would be to use empty layout for an element such as

`<foo>   </foo>`

I don't think there's an obviously right answer to this one. I have no objection to making the change.

>It is often the case that the "no, not that one" text node has &#x00A0; in it, rather than XML whitespace;

I would prefer to keep our definition of ignorable whitespace consistent with the other specs, for example the definition of element-only content in XSD and DTDs, the behaviour of xsl:strip-space, the SAX API, etc.

**Comment by graydon2014** [comment]
> I would prefer to keep our definition of ignorable whitespace consistent with the other specs, for example the definition of element-only content in XSD and DTDs, the behaviour of xsl:strip-space, the SAX API, etc.

How would you feel about optionally-a-higher-order-function, with the current default and an optional function parameter that gets used when present?

I think there's value to having the default stick to the existing XML-whitespace definition; I think there's also value in allowing the user to adjust the rules for whatever specific case they're handling. The pattern for that is "pass a function", though I'm not thinking of an existing case where the passed function is optional.

**Comment by michaelhkay** [comment]
>How would you feel about optionally-a-higher-order-function

I'm not sure how it would interact with the other options. The whole thing is in danger of becoming too complex. Don't forget that you can pre-process the XML, post-process the JSON, or do the whole thing by hand. This function doesn't have to do everything that anyone has ever dreamt of.

**Comment by michaelhkay** [comment]
I believe that, with application of PR #1658, the rules are now correct and consistent.

In selecting a layout, empty and empty-plus are chosen only if the elements are truly empty (no whitespace text nodes). However, if empty or empty-plus are explicitly chosen for elements that contain whitespace, the whitespace will be discarded.

Whitespace text nodes are ignored for list, list-plus, record, and sequence layouts.

I believe that all corresponds with common practice, so I think this issue should be closed by PR #1906.

**Comment by ChristianGruen** [comment]
Closed; see #1949 for updated feedback.
